tests/mock_network/messages.rs

```diff
@@ -73,69 +63,58 @@ fn send_and_receive() {
-                Some(Event::MessageReceived { content, src, .. }) => {
-                    request_received_count += 1;
-                    if req_content == content {
-                        let res_src = SrcLocation::Section(*node.our_prefix());
-                        let res_dst = match src {
-                            SrcLocation::Node(id) => DstLocation::Node(*id.name()),
-                            _ => panic!("Unexpected src location: {:?}", src),
-                        };
-
-                        if let Err(err) =
-                            node.inner
-                                .send_message(res_src, res_dst, res_content.clone())
-                        {
-                            trace!("Failed to send message: {:?}", err);
-                        }
-                        break;
-                    }
-                }
-                Some(_) => (),
-                _ => panic!("Event::MessageReceived not received"),
-            }
-        }
-    }
-
-    assert!(request_received_count >= quorum);
+    // For all expected recipients, poll until it receives the request message and send
+    // the response
+    let expected_req_recipients: Vec<_> = expected_recipients(&nodes, &req_dst).collect();
+    assert!(expected_req_recipients.len() >= quorum);
 
-    let _ = poll_all(&mut nodes);
+    for index in expected_req_recipients {
+        poll_until(&env, &mut nodes, |nodes| {
+            message_received(&nodes[index], &req_content)
+        });
 
-    let mut response_received_count = 0;
+        let res_src = SrcLocation::Section(*nodes[index].our_prefix());
+        let res_dst = DstLocation::Node(*nodes[sender_index].name());
 
-    loop {
-        match nodes[sender_index].try_recv_event() {
-            Some(Event::MessageReceived { content, .. }) => {
-                response_received_count += 1;
-                if res_content == content {
-                    break;
-                }
-            }
-            Some(_) => (),
-            _ => panic!("Event::MessageReceived not received"),
+        if let Err(err) = nodes[index]
+            .inner
+            .send_message(res_src, res_dst, res_content.clone())
+        {
+            trace!("Failed to send message: {:?}", err);
         }
     }
 
-    assert_eq!(response_received_count, 1);
+    // Poll until the response is delivered.
+    poll_until(&env, &mut nodes, |nodes| {
+        message_received(&nodes[sender_index], &res_content)
+    })
+}
+
+fn expected_recipients<'a>(
+    nodes: &'a [TestNode],
+    dst: &'a DstLocation,
+) -> impl Iterator<Item = usize> + 'a {
+    nodes
+        .iter()
+        .enumerate()
+        .filter(move |(_, node)| node.inner.is_elder() && node.inner.in_dst_location(dst))
+        .map(|(index, _)| index)
+}
+
+fn message_received(node: &TestNode, expected_content: &[u8]) -> bool {
+    if let Some(Event::MessageReceived { content, .. }) = node.try_recv_event() {
```

Comment:
just to ensure that membership change events (nodes joined or left) won't cause extra Event notification from this channel

Reply:
There might be extra events, but we don't care about them in this tests. I'll make it explicit.